static/app/components/modals/discoverAddToDashboardModal.tsx

```diff
@@ -0,0 +1,535 @@
+import set from 'lodash/set';
+
+import {validateWidget} from 'app/actionCreators/dashboards';
+import {addErrorMessage} from 'app/actionCreators/indicator';
+import {ModalRenderProps} from 'app/actionCreators/modal';
+import {Client} from 'app/api';
+import Button from 'app/components/button';
+import ButtonBar from 'app/components/buttonBar';
+import WidgetQueriesForm from 'app/components/dashboards/widgetQueriesForm';
+import SelectControl from 'app/components/forms/selectControl';
+import {PanelAlert} from 'app/components/panels';
+import {t} from 'app/locale';
+import space from 'app/styles/space';
+import {GlobalSelection, Organization, SelectValue, TagCollection} from 'app/types';
+import {
+  aggregateOutputType,
+  isAggregateField,
+  isLegalYAxisType,
+} from 'app/utils/discover/fields';
+import Measurements from 'app/utils/measurements/measurements';
+import withApi from 'app/utils/withApi';
+import withGlobalSelection from 'app/utils/withGlobalSelection';
+import withTags from 'app/utils/withTags';
+import {DISPLAY_TYPE_CHOICES} from 'app/views/dashboardsV2/data';
+import {
+  DashboardListItem,
+  DisplayType,
+  Widget,
+  WidgetQuery,
+} from 'app/views/dashboardsV2/types';
+import WidgetCard from 'app/views/dashboardsV2/widgetCard';
+import {generateFieldOptions} from 'app/views/eventsV2/utils';
+import Input from 'app/views/settings/components/forms/controls/input';
+import Field from 'app/views/settings/components/forms/field';
+
+export type DiscoverAddToDashboardModalOptions = {
+  organization: Organization;
+  defaultQuery?: string;
+};
+
+type Props = ModalRenderProps &
+  DiscoverAddToDashboardModalOptions & {
+    api: Client;
+    organization: Organization;
+    selection: GlobalSelection;
+    tags: TagCollection;
+  };
+
+type ValidationError = {
+  [key: string]: string[] | ValidationError[] | ValidationError;
+};
+
+type FlatValidationError = {
+  [key: string]: string | FlatValidationError[] | FlatValidationError;
+};
+
+type State = {
+  title: string;
+  displayType: Widget['displayType'];
+  interval: Widget['interval'];
+  queries: Widget['queries'];
+  loading: boolean;
+  errors?: Record<string, any>;
+  dashboards: SelectValue<string>[];
+  selectedDashboard?: SelectValue<string>;
+};
+
+const newQuery = {
+  name: '',
+  fields: ['count()'],
+  conditions: '',
+  orderby: '',
+};
+
+function mapErrors(
+  data: ValidationError,
+  update: FlatValidationError
+): FlatValidationError {
+  Object.keys(data).forEach((key: string) => {
+    const value = data[key];
+    // Recurse into nested objects.
+    if (Array.isArray(value) && typeof value[0] === 'string') {
+      update[key] = value[0];
+    } else if (Array.isArray(value) && typeof value[0] === 'object') {
+      update[key] = (value as ValidationError[]).map(item => mapErrors(item, {}));
+    } else {
+      update[key] = mapErrors(value as ValidationError, {});
+    }
+  });
+
+  return update;
+}
+
+function normalizeQueries(
```

Comment:
Same comment as above with [this](https://github.com/getsentry/sentry/blob/master/static/app/components/modals/addDashboardWidgetModal.tsx#L102) if possible